File: src/ffdraft/pipeline/market.py
"""The Phase-5 market pipeline: retained snapshot + published board -> arbitrage artifact.

**Boundary module.** This is the join point of the two halves of the product, and the join
is one-directional by construction:

* the **published tier artifact** supplies fair rank. Nothing here re-runs the model, so a
  market outage, a market bug or a cohort change can never cause an intrinsic rebuild;
* the **retained snapshot** supplies price. Nothing here touches a vendor, so an arbitrage
  board is reproducible from evidence months later;
* the result is written beside the intrinsic artifacts, and `build_metadata.json` is
  **merged**, never rewritten — Phase 4's tier-stability warning and every source it
  recorded survive intact (ADR-035, and the Phase-5 brief's requirement that the partial
  exit-gate warnings stay attached).
"""

from __future__ import annotations

import json
from collections.abc import Mapping, Sequence
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any

from ffdraft.arbitrage.build import build_arbitrage_records
from ffdraft.arbitrage.frozen import (
    ARBITRAGE_CONFIDENCE_VERSION,
    ARBITRAGE_METHOD_VERSION,
    ARBITRAGE_MODE,
)
from ffdraft.artifacts import record_schema_version, write_artifact, write_build_metadata
from ffdraft.config import AppConfig, load_app_config
from ffdraft.contracts import QualityCheck
from ffdraft.contracts.enums import MarketSignalType, Severity, SourceStatus
from ffdraft.market.cohorts import assignments_from_report
from ffdraft.market.current import build_current_market
from ffdraft.market.extra import load_extra_quotes
from ffdraft.market.history import RetainedHistory, load_trend_window
from ffdraft.market.snapshot import MarketSnapshotStore
from ffdraft.market.surface import build_surface_universe, coverage_checks
from ffdraft.market.trend import TREND_RULE, trend_series_records
from ffdraft.quality import QualityGate
from ffdraft.sources.ffc import FFC_SOURCE_ID
from ffdraft.sources.market import MFL_SOURCE_ID
from ffdraft.timeutil import isoformat_utc, utc_now

__all__ = ["ArbitrageBuildRequest", "ArbitrageBuildResponse", "run_arbitrage_build"]


@dataclass(frozen=True)
class ArbitrageBuildRequest:
    """Everything one arbitrage build needs. All of it is on disk."""

    season: int
    store: MarketSnapshotStore
    artifacts_dir: Path
    selection_path: Path
    snapshot_key: str | None = None
    as_of: datetime | None = None
    git_sha: str | None = None
    write: bool = True
    app: AppConfig | None = None
    #: The untruncated fair-ranked board `build-current` wrote. Without it the surface rule
    #: cannot run, because a player outside the published depth is absent from the tier
    #: artifact this stage reads — which is the whole reason the blind spot existed.
    full_board_path: Path | None = None
    #: Retained market sources beyond MyFantasyLeague to price this board with.
    extra_source_ids: tuple[str, ...] = (FFC_SOURCE_ID,)


@dataclass
class ArbitrageBuildResponse:
    """What the build produced."""

    build_id: str
    season: int
    snapshot_key: str
    arbitrage_mode: str
    method_version: str
    records: list[dict[str, Any]] = field(default_factory=list)
    coverage: dict[str, Any] = field(default_factory=dict)
    confidence_counts: dict[str, int] = field(default_factory=dict)
    unpriced_top_players: list[dict[str, Any]] = field(default_factory=list)
    trend_available: bool = False
    trend_history_keys: tuple[str, ...] = ()
    trend_series: list[dict[str, Any]] = field(default_factory=list)
    #: Which markets the published history actually covers. A source that priced the board
    #: but has no retained window is absent here, which is the fact the card needs.
    trend_series_sources: tuple[str, ...] = ()
    metadata: dict[str, Any] = field(default_factory=dict)
    written: list[Path] = field(default_factory=list)
    gate: QualityGate = field(default_factory=QualityGate)


def run_arbitrage_build(request: ArbitrageBuildRequest) -> ArbitrageBuildResponse:
    """Build and write the current arbitrage artifact from retained evidence."""
    settings = request.app or load_app_config()
    gate = QualityGate()
    stamped = (request.as_of or utc_now()).replace(microsecond=0)

    tiers = _read_json(request.artifacts_dir / "tiers.json")
    metadata = _read_json(request.artifacts_dir / "build_metadata.json")
    build_id = str(tiers["build_id"])
    tier_records = list(tiers["records"])

    selection = _read_json(request.selection_path)
    assignments = assignments_from_report(selection)
    gate.add(
        QualityCheck.ok(
            "arbitrage.cohort_selection",
            stage="arbitrage.pipeline",
            message=(
                f"cohort selection read from {request.selection_path.name} under rule "
                f"{selection.get('rule_version')} (ADR-039)"
            ),
            observed=f"{len(assignments)} preset assignment(s)",
        ),
    )

    key = request.snapshot_key or request.store.latest_key(MFL_SOURCE_ID, request.season)
    if key is None:
        gate.add(
            QualityCheck.fail(
                "arbitrage.no_retained_snapshot",
                stage="arbitrage.pipeline",
                message="no retained market snapshot; the Tier board is unaffected",
                observed=str(request.store.root),
                expected="at least one snapshot (ADR-038)",
            ),
        )
        return ArbitrageBuildResponse(
            build_id=build_id,
            season=request.season,
            snapshot_key="",
            arbitrage_mode=str(ARBITRAGE_MODE),
            method_version=ARBITRAGE_METHOD_VERSION,
            gate=gate,
        )

    snapshot = request.store.read(MFL_SOURCE_ID, request.season, key)
    history = load_trend_window(
        request.store,
        source_id=MFL_SOURCE_ID,
        season=request.season,
        now=snapshot.retrieved_at,
    )
    market = build_current_market(
        snapshot,
        assignments=assignments,
        now=stamped,
        history=history,
    )

    league_sizes = {
        preset_id: preset.teams for preset_id, preset in settings.league.presets.items()
    }
    league_sizes.update(
        {preset_id: preset.teams for preset_id, preset in settings.league.optional_presets.items()},
    )
    # The second market. Phase 10 built every part of this and connected none of it; a
    # published board that renders a market selector must actually carry more than one
    # market (ADR-067).
    extra = load_extra_quotes(
        request.store,
        season=request.season,
        source_ids=request.extra_source_ids,
        now=stamped,
        gate=gate,
        league_sizes=sorted(set(league_sizes.values())),
    )

    # The surface rule needs the whole board, so it runs only when `build-current` handed one
    # over. Absent, the build behaves exactly as it did before: the published prefix is the
    # surface, and nothing is rescued.
    full_board = _load_full_board(request.full_board_path, gate=gate)
    surfaces = {}
    surfaced_rows: list[dict[str, Any]] = []
    if full_board:
        published = {
            (str(row["league_preset_id"]), str(row["scoring_preset"]), str(row["player_id"]))
            for row in tier_records
        }
        blocks = sorted(
            {(str(row["league_preset_id"]), str(row["scoring_preset"])) for row in tier_records},
        )
        for preset_id, scoring in blocks:
            block_board = [
                row
                for row in full_board
                if str(row["league_preset_id"]) == preset_id
                and str(row["scoring_preset"]) == scoring
            ]
            universe = build_surface_universe(
                block_board,
                scoring_preset=scoring,
                league_preset_id=preset_id,
                memberships=extra.memberships,
                # `build-current --full-board` writes the untruncated universe, so an
                # absent market player here is genuinely unvalued rather than cut.
                board_is_complete=True,
            )
            surfaces[(preset_id, scoring)] = universe
            by_id = {str(row["player_id"]): row for row in block_board}
            surfaced_rows.extend(
                {**by_id[player_id], "league_preset_id": preset_id, "scoring_preset": scoring}
                for player_id in universe.exceptions
                if (preset_id, scoring, player_id) not in published and player_id in by_id
            )
        gate.extend(coverage_checks(list(surfaces.values())))

    result = build_arbitrage_records(
        tier_records,
        market=market,
        league_size_by_preset=league_sizes,
        build_id=build_id,
        season=request.season,
        generated_at=stamped,
        gate=gate,
        extra_quotes=extra.quotes or None,
        surfaces=surfaces or None,
        surfaced_rows=surfaced_rows,
    )

    response = ArbitrageBuildResponse(
        build_id=build_id,
        season=request.season,
        snapshot_key=key,
        arbitrage_mode=result.arbitrage_mode,
        method_version=result.method_version,
        records=result.records,
        coverage=result.coverage,
        confidence_counts=result.confidence_counts,
        unpriced_top_players=result.unpriced_top_players,
        trend_available=market.trend_available,
        trend_history_keys=market.trend_history_keys,
        gate=gate,
    )
    # Every retained ADP source contributes its own history, MyFantasyLeague included. Until
    # this was a list, it was one source's window passed by hand, and the second market's
    # chart could not exist however many snapshots the store held (ADR-081).
    response.trend_series = _trend_series(
        [market.retained, *(extra.histories[source] for source in sorted(extra.histories))],
        arbitrage_records=result.records,
        build_id=build_id,
        league_sizes=league_sizes,
    )
    response.trend_series_sources = tuple(
        sorted({str(record["market_source_id"]) for record in response.trend_series}),
    )

    response.metadata = _merge_metadata(
        metadata,
        response=response,
        market=market,
        snapshot_row_count=len(snapshot.rows),
        selection=selection,
        git_sha=request.git_sha,
        gate=gate,
        extra_sources=extra.sources,
    )

    if request.write and gate.passed:
        paths, checks = write_artifact(
            "arbitrage",
            result.records,
            out_dir=request.artifacts_dir,
            build_id=build_id,
            generated_at=_generated_at(tiers, stamped),
            arbitrage_mode=result.arbitrage_mode,
        )
        gate.extend(checks)
        response.written.extend(paths)
        if response.trend_series:
            # The chart's data, written from the retained snapshots the trend was computed
            # over. This is what lets a static page draw a history without the browser
            # calling a vendor (ADR-066). An empty series simply is not written: a young
            # store has nothing to draw, and an artifact of empty arrays would be bytes
            # shipped to say so.
            series_paths, series_checks = write_artifact(
                "market_trend_series",
                response.trend_series,
                out_dir=request.artifacts_dir,
                build_id=build_id,
                generated_at=_generated_at(tiers, stamped),
            )
            gate.extend(series_checks)
            response.written.extend(series_paths)
        metadata_paths, metadata_checks = write_build_metadata(
            response.metadata,
            out_dir=request.artifacts_dir,
        )
        gate.extend(metadata_checks)
        response.written.extend(metadata_paths)
    return response


def _trend_series(
    histories: Sequence[RetainedHistory | None],
    *,
    arbitrage_records: Sequence[Mapping[str, Any]],
    build_id: str,
    league_sizes: dict[str, int],
) -> list[dict[str, Any]]:
    """One history series per source, per player **that source priced**, per preset block.

    Scoped by the published arbitrage rows rather than by the tier board, and per source
    rather than per player. That is narrower than it first looks, and both halves matter:

    * *per player* keeps ADR-066's rule that a series for a card nobody can open is weight
      every visitor downloads for nothing — and, because the scope is the arbitrage row, a
      market-surfaced exception (ADR-063) keeps his chart;
    * *per source* is the half a single-market build could not have needed. The trailing
      window is seven days wide, so it holds players a source priced on Tuesday and dropped
      by Friday. Charting those would put a history under a market that has no current price
      for him — which the card would render as "no FFC ADP" beside an FFC line. Found by the
      cross-artifact check on a real 2026 build, on 20 players (ADR-081).

    **The artifact stays source-specific.** A player both markets price gets two records —
    one FFC, one MyFantasyLeague — carrying their own cohorts, their own points and their own
    slopes. There is deliberately no ``cross`` record: a synthetic cross-market ADP history
    would be a line no capture produced, and the cross view's job is to show the two real
    series beside each other rather than to average them away.
    """
    schema_version = record_schema_version("market_trend_series")
    priced: dict[tuple[str, str, str], set[str]] = {}
    for record in arbitrage_records:
        block = (str(record["league_preset_id"]), str(record["scoring_preset"]))
        for entry in record.get("markets") or ():
            if str(entry.get("market_signal_type")) != str(MarketSignalType.ADP):
                continue
            priced.setdefault((*block, str(entry["source_id"])), set()).add(
                str(record["player_id"]),
            )

    records: list[dict[str, Any]] = []
    for history in histories:
        if history is None or history.is_empty:
            continue
        for (preset_id, scoring, source_id), players in sorted(priced.items()):
            if source_id != history.source_id:
                continue
            league_size = league_sizes.get(preset_id)
            if league_size is None:
                continue
            cohort_id = history.cohort_for(scoring, league_size)
            if cohort_id is None:
                continue
            records.extend(
                trend_series_records(
                    history.observations_for(cohort_id),
                    trends=history.trends_for(cohort_id),
                    build_id=build_id,
                    market_source_id=history.source_id,
                    scoring_preset=scoring,
                    league_preset_id=preset_id,
                    cohort_id=cohort_id,
                    window_days=history.rule.window_days,
                    schema_version=schema_version,
                    players=players,
                ),
            )
    return records


def _generated_at(tiers: dict[str, Any], fallback: datetime) -> datetime:
    """The arbitrage envelope carries the tier build's generation time.

    Both artifacts describe one build. Stamping the arbitrage envelope with the moment the
    market layer happened to run would make two files from the same build disagree about
    when that build was.
    """
    from ffdraft.timeutil import parse_utc

    raw = tiers.get("generated_at_utc")
    return parse_utc(str(raw)) if raw else fallback


def _load_full_board(path: Path | None, *, gate: QualityGate) -> list[dict[str, Any]]:
    """The untruncated board `build-current` wrote, or an empty list with the reason.

    Missing is not an error. The fixture pipeline and every offline rebuild run without it,
    and the correct behaviour there is the pre-surface one — publish the prefix, rescue
    nobody — rather than a crash. What must never happen is a *silent* skip, so an absent or
    unreadable file is recorded either way.
    """
    if path is None:
        return []
    try:
        rows = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError) as error:
        gate.add(
            QualityCheck.fail(
                "arbitrage.full_board_unreadable",
                stage="arbitrage.pipeline",
                message=(
                    f"the untruncated board at {path} could not be read, so no player is "
                    f"surfaced from beyond the published depth: {error}"
                ),
                observed=type(error).__name__,
                expected="the board written by build-current --full-board",
                severity=Severity.WARNING,
            ),
        )
        return []
    return [dict(row) for row in rows]


def _merge_metadata(
    metadata: dict[str, Any],
    *,
    response: ArbitrageBuildResponse,
    market: Any,
    snapshot_row_count: int,
    selection: dict[str, Any],
    git_sha: str | None,
    gate: QualityGate,
    extra_sources: Sequence[Mapping[str, Any]] = (),
) -> dict[str, Any]:
    """Add the arbitrage block to the intrinsic build's metadata, keeping everything else.

    Phase 4 published a board that had failed its tier stability gate and recorded a warning
    saying so. An arbitrage build that overwrote this file would erase that warning, which
    is the exact failure the Phase-5 brief forbids. So the existing warnings, sources and
    quality-gate summary are preserved and only added to.
    """
    merged = dict(metadata)
    warnings = list(merged.get("warnings", ()))
    sources = [dict(source) for source in merged.get("sources", ())]

    market_source = {
        "source_id": market.source_id,
        "status": str(SourceStatus.WARNING if gate.warnings else SourceStatus.PASS),
        "retrieved_at_utc": isoformat_utc(market.snapshot_at_utc),
        # MFL publishes no data-as-of time and never will (docs/DATA_SOURCES.md 13.5).
        "source_as_of_utc": None,
        "record_count": snapshot_row_count,
        "warnings": sorted(
            {flag for price in market.prices.values() for flag in price.quality_flags},
        ),
    }
    sources = [source for source in sources if source.get("source_id") != market.source_id]
    sources.append(market_source)

    for check in gate.warnings:
        if check.message not in warnings:
            warnings.append(check.message)

    merged["arbitrage_mode"] = response.arbitrage_mode
    merged["arbitrage_model_version"] = None
    merged["arbitrage_method_version"] = response.method_version
    merged["sources"] = sorted(sources, key=lambda source: str(source["source_id"]))
    merged["warnings"] = warnings
    merged["market"] = {
        "source_id": market.source_id,
        "snapshot_key": response.snapshot_key,
        "snapshot_at_utc": isoformat_utc(market.snapshot_at_utc),
        "source_as_of_utc": None,
        "cohort_rule_version": str(selection.get("rule_version", "")),
        "cohort_report": str(selection.get("snapshot_key", "")),
        "confidence_rubric_version": ARBITRAGE_CONFIDENCE_VERSION,
        "trend_rule_version": TREND_RULE.version,
        "trend_available": response.trend_available,
        "trend_history_snapshots": len(response.trend_history_keys),
        # Per source, because "is there a chart?" is a per-source question and answering it
        # with MyFantasyLeague's number is exactly how a real FFC history went unnoticed
        # (ADR-081). A source listed here with `series: false` priced the board and has no
        # retained window; one with `trend_available: false` has a window whose span has not
        # yet reached the frozen rule, which is a different and temporary state.
        "trend_sources": [
            {
                "source_id": market.source_id,
                "snapshots": len(response.trend_history_keys),
                "trend_available": response.trend_available,
                "series": market.source_id in response.trend_series_sources,
            },
            *(
                {
                    "source_id": str(entry["source_id"]),
                    "snapshots": int(entry.get("trend_history_snapshots", 0)),
                    "trend_available": bool(entry.get("trend_available", False)),
                    "series": str(entry["source_id"]) in response.trend_series_sources,
                }
                for entry in extra_sources
                if entry.get("status") == "priced"
            ),
        ],
        "assignments": [
            {
                "scoring_preset": assignment.scoring_preset,
                "league_size": assignment.league_size,
                "cohort_id": assignment.cohort.cohort_id,
                "exact": assignment.exact,
                "sufficient": assignment.sufficient,
                "source_format_detail": assignment.source_format_detail,
                # The clauses that failed, verbatim ("total_drafts 125 < 300"). Published so a
                # consumer can say *why* every row on this board reads `low` confidence
                # without embedding today's measurement in its own source. Without it the
                # frontend would either hardcode a number that goes stale within a day or
                # show an unexplained label on 2,122 rows (ADR-041, ADR-045).
                "failed_clauses": list(assignment.failed_clauses),
            }
            for _, assignment in sorted(market.assignments.items())
        ],
        "coverage": response.coverage,
        "confidence_counts": response.confidence_counts,
        "unpriced_top_players": len(response.unpriced_top_players),
    }
    if git_sha:
        merged["git_sha"] = git_sha
    summary = gate.summary()
    existing = dict(merged.get("quality_gate", {}))
    merged["quality_gate"] = {
        "status": "fail"
        if summary["status"] == "fail" or existing.get("status") == "fail"
        else "pass",
        "critical_failures": int(existing.get("critical_failures", 0))
        + len(gate.critical_failures),
        "warnings": int(existing.get("warnings", 0)) + len(gate.warnings),
    }
    return merged


def _read_json(path: Path) -> dict[str, Any]:
    payload: dict[str, Any] = json.loads(path.read_text(encoding="utf-8"))
    return payload

File: web/src/data/market.ts
/**
 * What the market metadata means, derived rather than asserted.
 *
 * Every number in this file comes out of `build_metadata.json` or the arbitrage records. None
 * of it is written down in TypeScript, and that decision has now been tested by events: the
 * launch board was priced by 125 keeper-free drafts against a frozen bar of 300 and every row
 * read `low`; a week later the same frozen rule was clearing the same bar and most rows read
 * `medium`. A panel that had hardcoded either state would have been wrong within days
 * (ADR-045, ADR-052).
 *
 * So this module has no notion of a normal condition. `summarizeMarket` reports whatever the
 * rows say — one label or several — and `marketHeadline` turns that into a sentence for any
 * of them. The same components render `low`, `medium`, `high` and a mixed board, and a trend
 * that is null renders exactly as correctly as one that is not.
 *
 * The distinction the module exists to keep straight: **`confidence` is market-data quality,
 * not a probability** (ADR-041). A `low` row is not a player the model is unsure about. It is
 * a price with little draft evidence behind it.
 */

import type { ArbitrageRecord, BuildMetadata, Confidence, ScoringPreset } from "./contracts";

export interface CohortAssignmentView {
  readonly scoringPreset: ScoringPreset;
  readonly leagueSize: number;
  readonly cohortId: string;
  readonly exact: boolean;
  readonly sufficient: boolean;
  readonly sourceFormatDetail: string;
  readonly failedClauses: readonly string[];
}

export function cohortAssignment(
  metadata: BuildMetadata,
  scoring: ScoringPreset,
  teams: number,
): CohortAssignmentView | null {
  const match = metadata.market?.assignments?.find(
    (entry) => entry.scoring_preset === scoring && entry.league_size === teams,
  );
  if (match === undefined) return null;
  return {
    scoringPreset: match.scoring_preset,
    leagueSize: match.league_size,
    cohortId: match.cohort_id,
    exact: match.exact,
    sufficient: match.sufficient,
    sourceFormatDetail: match.source_format_detail,
    failedClauses: match.failed_clauses ?? [],
  };
}

/**
 * A frozen sufficiency clause, e.g. `total_drafts 214 < 300`, turned into a sentence.
 *
 * The clause strings are the build's own words. Parsing them keeps the observed value and
 * the bound out of this file; an unrecognized clause is passed through verbatim rather than
 * dropped, so a future rule version degrades to "readable" rather than "invisible".
 */
export function explainClause(clause: string): string {
  const parsed = /^(\w+)\s+(\S+)\s*<\s*(\S+)$/.exec(clause.trim());
  if (parsed === null) return clause;
  const [, field, observed, bound] = parsed;
  const templates: Readonly<Record<string, (a: string, b: string) => string>> = {
    total_drafts: (a, b) =>
      `the cohort holds ${a} drafts against the ${b} the rule requires`,
    priced_players: (a, b) => `only ${a} of the ${b} required players carry a price`,
    top100_board_coverage: (a, b) =>
      `${a} of the top 100 board players are priced, against ${b} required`,
    top150_board_coverage: (a, b) =>
      `${a} of the top 150 board players are priced, against ${b} required`,
    median_top150_sample_size: (a, b) =>
      `the median top-150 player was priced by ${a} drafts, against ${b} required`,
    identity_coverage: (a, b) => `identity coverage is ${a}, against ${b} required`,
  };
  const template = templates[field ?? ""];
  return template === undefined ? clause : template(observed ?? "?", bound ?? "?");
}

export interface MarketConditionSummary {
  /** Distinct confidence labels on the rows currently in scope. */
  readonly confidenceCounts: Readonly<Record<Confidence, number>>;
  /** The single label, when every row in scope carries the same one. Null on a mixed board. */
  readonly uniform: Confidence | null;
  /** The label the largest number of rows carry. Never null when there is a row. */
  readonly dominant: Confidence | null;
  readonly assignment: CohortAssignmentView | null;
  /** Median `market_sample_size` over the rows in scope. The direct per-player evidence. */
  readonly medianSampleSize: number | null;
  readonly rows: number;
  readonly trendAvailable: boolean;
  readonly trendSnapshots: number | null;
  readonly snapshotAtUtc: string | null;
  readonly sourceId: string;
}

const EMPTY_COUNTS: Readonly<Record<Confidence, number>> = {
  high: 0,
  medium: 0,
  low: 0,
  unknown: 0,
};

export function summarizeMarket(
  metadata: BuildMetadata,
  records: readonly ArbitrageRecord[],
  scoring: ScoringPreset,
  teams: number,
): MarketConditionSummary {
  const counts: Record<Confidence, number> = { ...EMPTY_COUNTS };
  const samples: number[] = [];
  for (const record of records) {
    counts[record.confidence] += 1;
    if (record.market_sample_size !== null) samples.push(record.market_sample_size);
  }
  const present = (Object.keys(counts) as Confidence[]).filter((key) => counts[key] > 0);
  let dominant: Confidence | null = null;
  for (const key of present) {
    if (dominant === null || counts[key] > counts[dominant]) dominant = key;
  }
  samples.sort((a, b) => a - b);
  const middle = Math.floor(samples.length / 2);
  const medianSampleSize =
    samples.length === 0
      ? null
      : samples.length % 2 === 1
        ? (samples[middle] ?? null)
        : ((samples[middle - 1] ?? 0) + (samples[middle] ?? 0)) / 2;

  return {
    confidenceCounts: counts,
    uniform: present.length === 1 ? (present[0] ?? null) : null,
    dominant,
    assignment: cohortAssignment(metadata, scoring, teams),
    medianSampleSize,
    rows: records.length,
    trendAvailable: metadata.market?.trend_available ?? false,
    trendSnapshots: metadata.market?.trend_history_snapshots ?? null,
    snapshotAtUtc: metadata.market?.snapshot_at_utc ?? null,
    sourceId: metadata.market?.source_id ?? "myfantasyleague_adp",
  };
}

/** The human name of a market source. Never "Consensus ADP": this product built no consensus. */
export function marketSourceLabel(sourceId: string): string {
  return sourceId === "myfantasyleague_adp" ? "MyFantasyLeague ADP" : sourceId;
}

export const CONFIDENCE_LABELS: Readonly<Record<Confidence, string>> = {
  high: "High market data",
  medium: "Medium market data",
  low: "Low market data",
  unknown: "No market sample",
};

export const CONFIDENCE_SHORT: Readonly<Record<Confidence, string>> = {
  high: "High",
  medium: "Medium",
  low: "Low",
  unknown: "Unknown",
};

/**
 * What the label is a statement about. Read on every surface that shows a confidence value,
 * because "low" beside a player's name reads as "the model is unsure about him" unless it is
 * told otherwise, and that is the opposite of what the field means.
 */
export const CONFIDENCE_MEANING =
  "Market-data quality: how much draft evidence stands behind this price. " +
  "It is not a probability that the player is a bargain, and it says nothing about the " +
  "intrinsic projection beside it.";

/** Trend semantics, in words, so a sign is never the only channel (ADR-042). */
export function describeTrend(trend: number | null): {
  readonly text: string;
  readonly direction: "earlier" | "later" | "flat" | "unknown";
} {
  if (trend === null) {
    return { text: "Trend collecting", direction: "unknown" };
  }
  if (trend === 0) return { text: "No measured movement", direction: "flat" };
  return trend > 0
    ? { text: "Moving earlier (more expensive)", direction: "earlier" }
    : { text: "Moving later (less expensive)", direction: "later" };
}

export const TREND_UNAVAILABLE_EXPLANATION =
  "A trend needs at least three observation days spanning three days in our own retained " +
  "snapshot store. Until then there is no estimate — which is not the same as no movement.";

/**
 * The bargain sentence, spelled out.
 *
 * `rank_gap = market_adp − fair_rank`. Positive means the market drafts him later than his
 * fair rank, so he can be had cheaply (ADR-040). Direction is stated in words as well as sign
 * so nothing depends on colour or on the reader knowing that smaller pick numbers are earlier.
 */
export function describeGap(rankGap: number): {
  readonly kind: "bargain" | "premium" | "even";
  readonly sentence: string;
  readonly compact: string;
} {
  const magnitude = Math.abs(rankGap).toFixed(1);
  if (Math.abs(rankGap) < 0.05) {
    return { kind: "even", sentence: "The market drafts him at his fair rank.", compact: "Even" };
  }
  if (rankGap > 0) {
    return {
      kind: "bargain",
      sentence: `The market drafts him ${magnitude} picks later than his fair rank.`,
      compact: `+${magnitude} picks later`,
    };
  }
  return {
    kind: "premium",
    sentence: `The market drafts him ${magnitude} picks earlier than his fair rank.`,
    compact: `\u2212${magnitude} picks earlier`,
  };
}

/**
 * Confidence labels, ordered weakest to strongest.
 *
 * Used to decide whether a board's market condition needs a warning tone or an informational
 * one. `unknown` sits below `low`: a row with no sample at all is weaker evidence than a row
 * with a small one.
 */
export const CONFIDENCE_ORDER: readonly Confidence[] = ["unknown", "low", "medium", "high"];

/** True when this label is weak enough that a board carrying it should say so up front. */
export function isWeakConfidence(confidence: Confidence): boolean {
  return confidence === "low" || confidence === "unknown";
}

export interface MarketHeadline {
  readonly tone: "info" | "warning";
  readonly sentence: string;
}

/**
 * The one-line market condition, for whatever the board actually is.
 *
 * Three shapes, all derived: every row carrying one label, a mixed board, and no rows at all.
 * There is no branch here that assumes a particular label, which is the whole point — the
 * product moved `low` -> `medium` on its own between two daily refreshes and this sentence
 * had to move with it without a code change.
 */
export function marketHeadline(summary: MarketConditionSummary): MarketHeadline | null {
  if (summary.rows === 0) return null;
  const { uniform, confidenceCounts, dominant } = summary;
  if (uniform !== null) {
    return {
      tone: isWeakConfidence(uniform) ? "warning" : "info",
      sentence: `Every priced row on this board carries ${CONFIDENCE_SHORT[
        uniform
      ].toLowerCase()} market-data confidence.`,
    };
  }
  const parts = CONFIDENCE_ORDER.filter((key) => confidenceCounts[key] > 0)
    .reverse()
    .map((key) => `${String(confidenceCounts[key])} ${CONFIDENCE_SHORT[key].toLowerCase()}`);
  return {
    tone: dominant !== null && isWeakConfidence(dominant) ? "warning" : "info",
    sentence: `Market-data confidence across these ${String(summary.rows)} priced rows: ${parts.join(", ")}.`,
  };
}
